In [ ]:
# Initialize Otter
import otter
grader = otter.Notebook("CT-06-Uebungen_Template.ipynb")

# Vorlesung "Computational Thinking"
### Woche 6 - Übungen
#### Prof. Dr.-Ing. Martin Hobelsberger

### Lernziele dieser Einheit

* Wiederholungen zu Python Grundlagen

## Übungen zum Warm-Up

***Aufgabe 1.*** Schreiben Sie eine Funktion ``index_at(text, char)``, welche die erste vorkommende Position eines gesuchten Zeichens ``char`` in einer Zeichenkette ``text`` zurückgibt. Wir beginnen bei 0 zu zählen, d.h. 

```python
index_at("abc", "a") == 0
```

In [1]:
def index_at(text, char):
    # BEGIN SOLUTION
    for i in range(len(text)):
        if char == text[i]:
            return i
    return None
    # END SOLUTION

In [ ]:
grader.check("q1")

In [8]:
text = 'May the force be with you!'
char = 'f'
index = index_at(text, char)
print(f'Der Buchstabe {zeichen} ist an Stelle {index} im Text: {text}')

Der Buchstabe f ist an Stelle 8 im Text: May the force be with you!


***Aufgabe 2:*** Schreiben Sie eine Yoda-Funktion ``yoda(text)``. Diese gibt einen übergebenen Satz ``text`` mit umgekehrten Wörtern wieder zurück.

```python
yoda('Ich bin bereit') == 'bereit bin Ich'
```

**Tipp:** Sehen Sie sich die String-Methoden ``join()`` und ``split()`` einer Zeichenkette ``str`` genauer an. 

In [16]:
def yoda(text):
    # BEGIN SOLUTION
    return ' '.join(text.split(' ')[::-1])
    # END SOLUTION

In [ ]:
grader.check("q2")

In [22]:
yoda('Hallo mein Name ist Hase')

'Hase ist Name mein Hallo'

***Aufgabe 3.*** Schreiben Sie ein Programm, das eine Liste (mit einer belibigen Anzahl an numerischen Elementen z.B.,

```python
numbers = [2, 42, 23.2, 34, 22332.5]
```

als Eingabe entgegen nimmt und alle Elemente der Liste mit einem **geraden Index** zu einem Summenwert sowie alle Elemente mit **ungeraden Index** zu einem Summenwert addiert. Es sollen beide Summenwerte ausgegeben werden.

In [26]:
# BEGIN SOLUTION
numbers = [1, 2, 3, 4, 5, 6]
even_sum = sum(numbers[0::2])
odd_sum = sum(numbers[1::2])
print((even_sum, odd_sum))
# END SOLUTION

(9, 12)


***Aufgabe 4.*** Schreiben Sie eine Funktion ``mask(numbers)``, die eine Liste (mit ganzen Zahlen als Elemente) entgegen nimmt, für jedes Element der Liste überprüft, ob es gerade oder ungerade ist und eine neue Liste mit den entsprechenden String-Elementen ``'gerade'`` bzw. ``'ungerade'`` zurückliefert.

Zum Beispiel sollte gelten:

```python
numbers = [7,34,2,22,31]
mask(numbers) == ['ungerade', 'gerade', 'gerade', 'gerade', 'ungerade']
```

In [32]:
def mask(numbers):
    # BEGIN SOLUTION
    return ['gerade' if x % 2 == 0 else 'ungerade' for x in numbers]
    # END SOLUTION

In [ ]:
grader.check("q4")

In [33]:
numbers = [7, 34, 2, 22, 31]
mask(numbers)

['ungerade', 'gerade', 'gerade', 'gerade', 'ungerade']

***Aufgabe 5.*** Schreiben Sie eine Funktion ``median(numbers)``, das eine Liste (mit einer belibigen Anzahl ``n`` an ganzzahligen Elementen; z.B. ``numbers = [2, 42, 232, 34, 5]``) als Eingabe entgegen nimmt, den [Median](https://de.wikipedia.org/wiki/Median) der Listenelemente berechnet und ausgibt.

**Tipp:** Mit der Methode ``sort`` einer Liste ``list`` könen Sie eine entsprechende Liste aus Zahlen sortieren.

In [37]:
def median(numbers):
    # BEGIN SOLUTION
    numbers_copy = numbers.copy()
    numbers_copy.sort()
    length = len(numbers_copy)
    if length == 0:
        return 0
    elif length % 2 == 1:
        return numbers_copy[length//2]
    else:
        return (numbers_copy[length//2] + numbers_copy[(length-1)//2]) / 2.0
    # END SOLUTION

In [ ]:
grader.check("q5")

***Aufgabe 6.*** Schreiben Sie eine Funktion die eine Liste (mit ganzen Zahlen als Elemente) entgegen nimmt und darüber die Quersumme bildet. Die Funktion soll auf der Konsole den Wert der Quersumme, die Anzahl der Elemente und ob die Quersumme gerade oder ungerade ist ausgeben. Als Rückgabewert wird die Quersumme geliefert. 

In [45]:
# BEGIN SOLUTION
def sum_numbers(numbers):
    length = len(numbers)
    avg_sum = sum(numbers)
    print(f'sum: {avg_sum}, n_elements: {length}, even: {avg_sum%2==0}')
    return avg_sum
# END SOLUTION

***Aufgabe 7. (Persönliche Daten aus Zeichenketten ausgeben)*** In einer Textdatei sind Zeile für Zeile Angaben zu einer Person gespeichert:

`'Name: Dörte Weber;'`

`'Adresse: Kobelweg 12, 86391 Stadtbergen;'`

`'Alter: 66;'`

Die Zeichenketten dieser Datei werden zeilenweise eingelesen in die Variable `row`. Schreiben Sie einen Code, der mit `print` nur die persönlichen Daten ausgibt (d.h. den String hinter dem Doppelpunkt und ohne Leerzeichen beginnend).

Hier im Beispiel, muss also aus `row` der gewünschte String extrahiert werden: 
```'Dörte Weber;'```. Teste dein Programm auch für die anderen Zeilen.

**Tipp:** Eventuell hilft Ihnen ``index_at`` von Aufgabe 1 und die Methode ``strip`` der Klasse ``str``.

In [51]:
row = "Name: Dörte Weber;"

In [57]:
# BEGIN SOLUTION
row[index_at(row, ':')+1::].strip()
# END SOLUTION

'Dörte Weber;'

***Aufgabe 8. (Kürzen von Zeichenketten)*** Für einen Fashion-Blog soll eine kurze Vorschau der Artikel im Umfang von 35 Zeichen angezeigt werden.
Der Text soll dabei mit drei Punkten `...` enden, um anzuzeigen, dass es noch mehr Inhalt gibt.

Schreiben Sie den Code, um einen `post` wie beschrieben neu zu formatieren und auszugeben. Für

In [59]:
post = "Die Herbst- und Winter-Saison ist eingeleitet und die Duftkerzen sind bereits angezündet. Jetzt fehlen also nur noch ein paar Parfums, die perfekt in die kältere Jahreszeit passen (aber in meinem Fall eigentlich auch das ganze Jahr über) – hier sind also meine aktuellen Lieblinge!"

Ist die gewünschte Ausgabe gleich

```python
Die Herbst- und Winter-Saison ist e...
```

In [63]:
length = len(post)
n = 35
if length < n:
    print(post)
else:
    print(f'{post[:n]}...')

Die Herbst- und Winter-Saison ist e...


***Aufgabe 9. (Verbessertes Kürzen von Zeichenketten)*** Können Sie Aufgabe 8 so lösen, dass die drei Punkte nach einem Leerzeichen eingefügt werden, um zu verhindern, dass Texte mitten in einem Wort abgeschnitten werden?

Ausgabe: 
```Die Herbst- und Winter-Saison ist...```

In [67]:
# BEGIN SOLUTION
length = len(post)
n = 35
if length < n:
    print(post)
else:
    words = post.split(' ')
    n_chars = 0
    index = 0
    while n_chars + len(words[index]) < n:
        n_chars += len(words[index])
        index += 1
    text = ' '.join(words[:index])
    print(f'{text}...')
# END SOLUTION

Die Herbst- und Winter-Saison ist...


In [64]:
"abc".find('b')

1

***Aufgabe 10. (Glückspiel)*** In ``w`` sei der Wert eines Würfelwerts gespeichert. Der Würfel kann die Werte 1, 2, 3, 4, 5 und 6 annehmen - Zum Beispiel gelte ``w==4``.

Schreibe Programmcode, der mit Hilfe des Modulo-Operators prüft, ob der Wert von ``w`` gerade oder ungerade ist. Das Programm soll z.B. ```3 ist ungerade``` oder ```4 ist gerade``` ausgeben.

**Beispielausgabe:**

```4 ist gerade```

In [74]:
w = 4

In [73]:
result = 'gerade' if w % 2 == 0 else 'ungerade'
print(f'{w} ist {result}')

5 ist ungerade


***Aufgabe 11. (Glückspiel 2)*** Nun wir der Würfen zweimal hintereinander geworfen. Wenn beide Würfelwerte ungerade sind, erhält der Spieler eine Gewinnsumme von 10 Euro. Wenn nur ein Würfelwert ungerade ist, beträgt die Gewinnsumme 0 Euro. Wenn beide Würfelwerte gerade sind, macht der Spieler einen Verlust von 10 Euro.

Ihre Aufgabe ist es, ein Programm zu schreiben, welches die Gewinnsumme bzw. den Verlust ausgibt.

**Beispielausgabe für w1=1 und w2=3:**
```Gewinnsumme beträgt 10 Euro.```

**Beispielausgabe für w1=1 und w2=4:**
```Gewinnsumme beträgt 0 Euro.``` 

**Beispielausgabe für w1=2 und w2=4:**
```Verlust beträgt 10 Euro.``` 

Testen Sie Ihren Programmcode für die verschiedenen Würfelwerte.

In [85]:
w1 = 1
w2 = 3

In [86]:
# BEGIN SOLUTION
outputs = ['Gewinnsumme beträgt 10 Euro', 'Gewinnsumme beträgt 0 Euro.', 'Verlust beträgt 10 Euro.']
if w1 % 2 == 1 and w2 % 2 == 1:
    print(outputs[0])
elif w1 % 2 == 0 and w2 % 2 == 0:
    print(outputs[2])
else:
    print(outputs[1])
# END SOLUTION

Gewinnsumme beträgt 10 Euro


***Aufgabe 12. (Versandkosten)*** Für einen Food-Online-Shop sollen Sie ein Programm schreiben, dass die Versandkosten berechnet. Die Variable ``bill`` enthält die Summe der Preise aller gekauften Waren. Die Versandkosten betragen ``8.95`` Euro. Ab einem Einkaufswert von ``40`` Euro betragen die Versandkosten nur ``5.95`` Euro; ab einem Einkaufswert von ``50`` Euro werden keine Versandkosten berechnet.

Deine Aufgabe ist es, ein Programm zu schreiben, dass die Versandkosten zum Preis (also zu der Zwischensumme) addiert und den (neuen) Gesamtpreis ausgibt. Weiterhin soll das Programm ggf. ausgeben, für wie viel Geld noch Waren gekauft werden müssen, damit die Bestellung versandkostenfrei geliefert wird bzw. nur 5,95 Euro beträgt.

**Hinweis**: Nutze die Funktion `round()`. Der zweite Parameter für die Funktion `round()` gibt an, auf wie viele Nachkommastellen gerundet werden soll. Bsp: `round(1.2934,2)` ergibt `1.29`. Im Zusammenhang mit der Ausgabe von Geldbeträgen kann diese Funktion sehr sinnvoll sein.

**Beispielausgabe für die Zwischensumme von 47.60 Euro:**

``Wenn Sie weitere Waren im Wert von 2.4 Euro einkaufen, erfolgt der Versand kostenfrei.
Gesamtpreis: 53.55 Euro``

**Beispielausgabe für die Zwischensumme von 36.25 Euro:**

``Wenn Sie weitere Waren im Wert von 3.75 Euro einkaufst, zahlen Sie für die Versandkosten nur 5.95 Euro. Wenn Sie weitere Waren im Wert von 13.75 Euro einkaufst, erfolgt der Versand kostenfrei.
Gesamtpreis: 45.2 Euro``

In [100]:
def print_shipping_message(bill):
    # BEGIN SOLUTION
    text = ''
    shipping = 0
    level = 0
    if bill < 40:
        shipping = 8.95
        level = 40
    elif bill < 50:
        shipping = 5.95
        level = 50

    if bill < 40:
        text = text + f' Wenn Sie weitere Waren im Wert von {round(40-bill, 2)} Euro einkaufen, zahlen Sie für die Versandkosten nur {5.95} Euro.'    

    if bill < 50:
        text = text + f' Wenn Sie weitere Waren im Wert von {round(50-bill,2)} Euro einkaufen, erfolgt der Versand kostenfrei.'

    text = text + f'\nGesamtpreis: {round(bill+shipping, 2)} Euro'
    print(text.strip())
    # END SOLUTION

In [101]:
bill = 47.60
print_shipping_message(bill)

Wenn Sie weitere Waren im Wert von 2.4 Euro einkaufen, erfolgt der Versand kostenfrei.
Gesamtpreis: 53.55 Euro


In [ ]:
bill = 36.25
print_shipping_message(bill)

***Aufgabe 13. (Öffnungszeiten)*** Schreiben Sie eine Funktion ``greetings(sunday, holiday)``, die Kunden eines Geschäfts grüßt.

1. Handelt es sich beim aktuellen Tag um einen Sonntag, d.h. ``sunday==True`` so lautet die Ausgabe:
``Wir haben geschlossen. Einen schönen Sonntag.``.
2. Sofern es sich nur um einen Feiertag handelt, d.h. ``holiday==True`` und ``sunday==False`` so lautet die Ausgabe:
``Wir haben geschlossen.``.
3. Sonst lautet die Ausgabe:
``Willkommen!``.


**Ausgabe für Sonntag:**
```Wir haben geschlossen. Einen schönen Sonntag.```

**Ausgabe für einen Feiertag:**
```Wir haben geschlossen.```

**Ausgbe für einen Werktag:**
```Willkommen!```


Mache dir Gedanken dazu, inwiefern die Reihenfolge der Bedingungen eine Rolle spielt!

In [102]:
def greetings(sunday, holiday):
    text = ''
    # BEGIN SOLUTION
    if holiday or sunday:
        text += 'Wir haben geschlossen.'
        if sunday:
            text += ' Einen schönen Sonntag.'
    else:
        text = 'Willkommen!'
    # END SOLUTION
    print(text)  

In [104]:
greetings(True, True)
greetings(True, False)
greetings(False, True)
greetings(True, False)
greetings(False, False)

Wir haben geschlossen. Einen schönen Sonntag.
Wir haben geschlossen. Einen schönen Sonntag.
Wir haben geschlossen.
Wir haben geschlossen. Einen schönen Sonntag.
Willkommen!


***Aufgabe 14. (Häufigkeitsanalyse)*** Schreiben Sie eine Funktion ``count_words(text, pattern)``, die die Anzahl der Vorkommnisse der Zeichenfolge ``pattern`` im Text ``text`` berechnet und zurückliefert. Dabei soll die Groß- und Kleinschreibweise ignoriert werden. Hierfür kannst du die `.lower()`-Funktion verwenden.

Öffne anschließend den text von TopicModellingPaper18, und ermittle, wie oft die Wörter "topic", "word" und "frequency" in diesem Paper vorkommen.

**Hinweis:** Die Referenzen, Angaben zu den Autoren, Überschriften zählen wir jetzt mit dazu. In bestimmten Analyseverfahren werden diese Textelemente häufig ausgeschlossen.

**Hinweis:** Mit `.count('data')` wird ausgegeben, wie oft der Teilstring `data` in einem String vorkommt. Wichtig: Diese Funktion unterscheidet zwischen Groß- und Kleinschreibung:

In [128]:
def count_words (text, pattern):
    # BEGIN SOLUTION
    return text.lower().count(pattern.lower())
    # END SOLUTION
    
with open("TopicModellingPaper18.txt", encoding = "utf-8") as file:
    content = file.read()
    # BEGIN SOLUTION
    patterns = ['topic', 'word', 'frequency']
    for pattern in patterns: 
        print(f'{count_words(content, pattern)}-mal kommt {pattern} vor.')
    # END SOLUTION

190-mal kommt topic vor.
91-mal kommt word vor.
2-mal kommt frequency vor.


In [ ]:
#Beispiel mit .count()
text = "Data data-science dataanalyst data"
print(text.count("data"))

Die Ausgabe Ihrer Analyse sollte so aussehen:

``190-mal kommt 'topic' vor.
  91-mal kommt 'word' vor.
  2-mal kommt 'frequency' vor.``

**Hinweis:** Falls noch nicht gemacht, nutze die `for-Schleife`, um die einzelnen Wörter zu analysieren. 

***Weitere Möglichkeiten:***
Ausgabe der Analyse in eine eigene Datei schreiben oder die Liste an Wörtern, die analysiert werden sollen, beliebig erweitern lassen, indem Wörter individuell eingegeben werden können. Experimentieren Sie selbst ein wenig herum!

***Aufgabe 15. (Häufigkeitsanalyse aller Wörter)*** Schreiben Sie eine Funktion ``count_all_words(text)``, welche ein Wörterbuch ``dict`` zurückliefert wobei ``key`` ein Wort in ``text`` ist und dessen Wert ``value`` die Anzahl der Vorkommnisse in ``text`` ist. Verwenden Sie erneut die Datei "TopicModellingPaper18.txt" zum Testen.

**Beachte:** Der Text enthält natürlich viele Satzzeichen. Sie müssen daher den Text zuerst an einem Leerzeichen trennen, und anschließend die ``word_cleaner`` Funktion (die bereits vorgegeben ist) verwenden, um Satzzeichen aus einem Wort zu entfernen. Diese Funktion entfernt alle Satzzeichen aus einem Wort, und gibt das Wort in Kleinbuchstaben zurück. Versuchen Sie als Erstes, diese Funktion nachzuvollziehen!
Schreiben Sie daraufhin Ihre Funktion ``count_all_words(text)``. Wir ignorieren alle Wörter die Sonderzeichen enthalten, z.B. ``'wouldn't'`` oder ``'http:\\www.hm.cs.edu'``.

**Hinweis 1:** Die Referenzen, Angaben zu den Autoren, Überschriften zählen wir mit dazu. 

**Hinweis 2:** Beachte, dass sich der leere String in deine Wörter einschleichen kann. Dieser zählt nicht als Wort!

In [133]:
def count_words(text, pattern):
    return text.lower().count(pattern.lower())

def word_cleaner(word):
    allowed_characters = "abcdefghijklmnopqrstuvwxyzüöäß"
    w = ""
    for character in word.lower():
        if character in allowed_characters:
            w += character
    return w
        
def count_all_words(text):
    # BEGIN SOLUTION
    word_count = {}
    words = set(map(lambda word: word_cleaner(word), text.split()))
    words -= {'', ' '}
    for word in words:
        count = count_words(text, word)
        if count > 0:
            word_count[word] = count
    return word_count
    # END SOLUTION
    
with open("TopicModellingPaper18.txt", encoding="utf-8") as file:
    content = file.read()
    # BEGIN SOLUTION
    word_count = count_all_words(content)
    for word, count in word_count.items():
        print(f'{count}-mal kommt {word} vor.')
    # END SOLUTION

40-mal kommt were vor.
1-mal kommt helped vor.
8-mal kommt jorgensen vor.
2-mal kommt settings vor.
1-mal kommt areas vor.
1-mal kommt steier vor.
1-mal kommt applications vor.
1-mal kommt doctoral vor.
1-mal kommt dealt vor.
1-mal kommt conceptions vor.
4-mal kommt often vor.
8-mal kommt belief vor.
2-mal kommt randomized vor.
2-mal kommt compared vor.
2-mal kommt expense vor.
1-mal kommt theoretically vor.
1-mal kommt criteria vor.
1-mal kommt einstein vor.
54-mal kommt proportion vor.
30-mal kommt low vor.
2-mal kommt computational vor.
1-mal kommt board vor.
1-mal kommt remarkably vor.
1-mal kommt shaped vor.
3-mal kommt assign vor.
1-mal kommt shopping vor.
8512-mal kommt t vor.
33-mal kommt turn vor.
37-mal kommt analysis vor.
2-mal kommt knipping vor.
2-mal kommt roberts vor.
59-mal kommt als vor.
4-mal kommt survey vor.
2-mal kommt arzarello vor.
2-mal kommt cases vor.
1-mal kommt yields vor.
1-mal kommt inference vor.
3-mal kommt seven vor.
1-mal kommt befitted vor.
2-mal komm

1-mal kommt events vor.
2-mal kommt bibliometric vor.
1-mal kommt inserted vor.
19-mal kommt reasoning vor.
2-mal kommt generalization vor.
37-mal kommt researcher vor.
2-mal kommt roh vor.
11-mal kommt didactical vor.
1-mal kommt advisory vor.
3-mal kommt instructional vor.
1-mal kommt matthew vor.
2-mal kommt planet vor.
1-mal kommt fidelity vor.
1-mal kommt outline vor.
1-mal kommt below vor.
9-mal kommt challenge vor.
3-mal kommt nearly vor.
4-mal kommt suit vor.
4-mal kommt upon vor.
3-mal kommt decide vor.
10-mal kommt conduct vor.
7-mal kommt reported vor.
8-mal kommt arithmetic vor.
4-mal kommt volume vor.
279-mal kommt so vor.
1-mal kommt constructs vor.
113-mal kommt from vor.
1-mal kommt chose vor.
2-mal kommt zuza vor.
3-mal kommt affect vor.
8-mal kommt position vor.
2-mal kommt attempted vor.
1-mal kommt classrooms vor.
2-mal kommt battista vor.
4-mal kommt belts vor.
1-mal kommt unify vor.
1-mal kommt abingdon vor.
37-mal kommt jrme vor.
1-mal kommt challenged vor.
1-mal

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit.

In [ ]:
grader.export(pdf=False, force_save=True)